### ISSUE TITLE
Missing params.best

### ISSUE THREAD

**Issue by konrad-roze** [awslabs/sockeye#234]
If you set low _max_num_epochs_ and _train_early_stop_when_ values then _params.best_ file is often missing although the training has succeeded. This is likely caused by the fact that there was no improvement of the model.
The params.best file should always be present after a model is trained.

**Comment by fhieber** [comment]
This can happen if you there not a single checkpoint made during training, independent whether there is any improvement. 
How many updates are performed in these cases? Can you set the `--checkpoint-frequency` to <= that number?

We can think about always doing at least one checkpoint when exiting the training loop.

**Comment by konrad-roze** [comment]
Decreasing checkpoint frequency in connection with reducing batch size worked for me as a workaround, thanks.
Still you probably need to handle this corner case differently. If you're not planning a fix, feel free to close this issue.

**Comment by tdomhan** [comment]
Could you share the exact CLI parameters and the size of your dataset?

Is it the case that during training the optimized metric did not improve a single time? I would probably make sense to handle this corner case. In any case something is probably fishy if your validation metric does not improve a single time during training. Are you dealing with a very small data set?

**Comment by konrad-roze** [comment]
This is very degenerate case that I cooked up for workflow testing/development. It turned out that my batch size was 2000 but I had only 1000 (or less) lines of input. Consequently there was no single checkpoint during the training. Still I believe that params.best should be produced and contain any/initial parameters.

**Comment by tdomhan** [comment]
I agree that a model should always be valid. With respect to testing it might make sense to take a look at the system copy test we use here:
https://github.com/awslabs/sockeye/blob/master/test/system/test_seq_copy_sys.py 
The test creates sequences of digits and we train small models to copy these sequences. The learning of these can be done in only a few updates.

**Comment by tdomhan** [comment]
This PR should fix the issue by writing out the initial parameters: https://github.com/awslabs/sockeye/pull/269 Let us know if that works for you.

**Comment by fhieber** [comment]
I am closing this for now, feel free to reopen if #269 did not solve the problem.
